### ISSUE TITLE
Best practices for handling 401 

### ISSUE THREAD

**Issue by andreliem** [nuxt-community/axios-module#25]
I've been looking at the best way of handling 401 with my static generated nuxt app and could use some guidance here.

I'm using Nuxt 1.0.0-rc8 and @nuxtjs/axios 4.0.0

I thought the best way to handle any unauthorized requests would be using a global catch by adding the following into `nuxt.config`

```
     responseInterceptor: (response, { store }) => {
        if (response.status === 401) {
          store.dispatch('logoutUser')
        }
        return response
      },
```

So anytime the API returns a 401, the site will dispatch the action `logoutUser`.

For some reason though, this is not triggered when I make calls to axios which return a 401.

Should I try something else?

Also, on the first page load I ideally want to run one auth check against the API to check the validity of a token as well.

I thought about adding this into the middleware, running an axios request and then setting the store before everything is setup. This doesn't seem to work either.

Any help would be greatly appreciated!

**Comment by andreliem** [comment]
I also tried using the `errorHandler` function to detect 401, which is working... but it seems like the `store` is not accessible from this function.

**Comment by andreliem** [comment]
Oh actually using middleware on its own without interceptors proved to be a good solution for me.

In case anyone is curious, I'm trying to set the session state of a user in the store before components load as I want specific components to hide/show content based on the store. Just redirecting to a login page isn't the UX I was looking to do.

```
export default async function ({ store }) {
    const authenticatedUser = await store.dispatch('authCheck')
    if (authenticatedUser) {
      store.dispatch('loginUser', authenticatedUser)
    } else {
      store.dispatch('logoutUser')
    }
}
```

**Comment by pi0** [comment]
@andreliem Thanks for your suggestion actually context was available using `this` (Like this.store) but it was inconsistent with other use cases, so I've made context available as second argument (In 4.2.1) ([docs](https://github.com/nuxt-community/axios-module#errorhandler)). 

Anyway, i agree with you. writing logic in the config is not a good idea, using middleware (Or plugins) is much better.

**Comment by andreliem** [comment]
Thanks @pi0. Good to know there is the context available for 4.2.1 in the error handler, I actually think it's a pretty decent location to have logic which dispatches actions to the store when errors like session expiry occur... as opposed to having each component handle the error or through some global mixin perhaps.

**Comment by utechs** [comment]
`export default function ({ $axios, store, redirect }) {
  $axios.onError((error) => {
    const code = parseInt(error.response && error.response.status)
    if (code === 503) {
      redirect('/maintenance')
    }
    if (code === 403) {
      redirect('/404')
    }
    if (code === 401) {
      store.$auth.logOut()
    }
  })
}`

I added this in axios plugin work fine looking for better practice
